src/main/scala/com/daml/projection/scaladsl/BatchSource.scala

```diff
@@ -186,4 +205,59 @@ object BatchSource {
       Consumer.treeEventSource(clientSettings, updatedProjection)
     }
   }
+
+  @FunctionalInterface
+  trait GetContractTypeId[E] {
+    def fromEvent(event: E): Option[Identifier]
+
+    def toJava: JGetContractTypeId[E] = (event: E) =>
+      fromEvent(event).map(i => J.Identifier.fromProto(toJavaProto(i))).toJava
+  }
+
+  object GetContractTypeId {
+    implicit val `from event`: GetContractTypeId[Event] = fromEvent
+    def fromEvent: GetContractTypeId[Event] = {
+      case Event(Created(createdEvent))   => Some(createdEvent.getTemplateId)
+      case Event(Archived(archivedEvent)) => Some(archivedEvent.getTemplateId)
+      case Event(Empty)                   => None
+    }
+
+    implicit val `from created event`: GetContractTypeId[CreatedEvent] = fromCreatedEvent
+    def fromCreatedEvent: GetContractTypeId[CreatedEvent] = (createdEvent: CreatedEvent) => createdEvent.templateId
+
+    implicit val `from archived event`: GetContractTypeId[ArchivedEvent] = fromArchivedEvent
+    def fromArchivedEvent: GetContractTypeId[ArchivedEvent] = (archivedEvent: ArchivedEvent) => archivedEvent.templateId
+
+    implicit val `from exercised event`: GetContractTypeId[ExercisedEvent] = fromExercisedEvent
+    def fromExercisedEvent: GetContractTypeId[ExercisedEvent] =
+      (exercisedEvent: ExercisedEvent) => exercisedEvent.templateId
+  }
+
+  @FunctionalInterface
```

Comment:
This is not intended to be used from Java, so no need for the attribute (on pure Scala API, if you create something that has both Scala and Java API usage it should be used, and if you provide a pure Java API)